{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b43e5dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "from gradientlab.data_utils.experiment_path import get_ckpt_path_by_exp_name\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import torch\n",
    "import torch.nn.functional as F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b33b0056",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============================================================\n",
      "Residual Scaling: 0.144338 (1/sqrt(2×24))\n",
      "Scaled 48 parameters:\n",
      "  ✓ decoder.blocks.0.attn.out_proj.weight\n",
      "  ✓ decoder.blocks.0.ffn.w3.weight\n",
      "  ✓ decoder.blocks.1.attn.out_proj.weight\n",
      "  ✓ decoder.blocks.1.ffn.w3.weight\n",
      "  ✓ decoder.blocks.2.attn.out_proj.weight\n",
      "  ✓ decoder.blocks.2.ffn.w3.weight\n",
      "  ✓ decoder.blocks.3.attn.out_proj.weight\n",
      "  ✓ decoder.blocks.3.ffn.w3.weight\n",
      "  ✓ decoder.blocks.4.attn.out_proj.weight\n",
      "  ✓ decoder.blocks.4.ffn.w3.weight\n",
      "  ... and 38 more\n",
      "============================================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "exp_name = Path(\".\").resolve().absolute().name\n",
    "ckpt_path = get_ckpt_path_by_exp_name(exp_name)\n",
    "\n",
    "device = \"cuda\"\n",
    "model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True).eval().to(device)\n",
    "tokenizer = AutoTokenizer.from_pretrained(ckpt_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "21e7ac79",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_orig = pd.read_csv(\"/home/mascit/Downloads/dataset_Hanna_updated_preprocessed.csv\")\n",
    "df = df_orig[[\"sgRNA_23bp\", \"BE39:MELJUSO:zscore\"]].copy()\n",
    "df = df.rename({\"sgRNA_23bp\": \"sequence\", \"BE39:MELJUSO:zscore\": \"target\"}, axis=1)\n",
    "df = df[~df[\"target\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5a8b11b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"target\"] = df[\"target\"].apply(lambda x: 1 if x < -2 else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99973400",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"target\"] = df[\"target\"].apply(lambda x: round(x, 1))\n",
    "df[\"target\"] = df[\"target\"].apply(lambda x: x if x >= -4 else -4.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c9b0e327",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12032"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = df.to_dict(orient=\"records\")\n",
    "len_data = len(data)\n",
    "len_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "983689e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sgRNA sequence', 'A_Content', 'T_Content', 'C_Content', 'G_Content',\n",
       "       'GC_Content', 'Tm_Global', 'Tm_PAM_Proximal', 'Tm_Middle', 'Stem_Loop',\n",
       "       'Free_Energy', 'Base_Accessibility', 'sgRNA context sequence',\n",
       "       'Gene symbol', 'Ensembl gene ID', 'Ensembl transcript ID',\n",
       "       'Gene strand', 'Genome assembly', 'Chromosome',\n",
       "       'sgRNA genomic position', 'sgRNA strand', 'PAM', 'Edit', '# edits',\n",
       "       '# silent edits', 'Nucleotide edits', 'Amino acid edits',\n",
       "       'Mutation category', 'Mutation bin', 'Clinical significance',\n",
       "       'On-target efficacy score', 'Match Bin I counts', 'offtargets-filter',\n",
       "       'BE39:MELJUSO:zscore', 'CAS9:MELJUSO:zscore', 'BE39:A375:zscore',\n",
       "       'CAS9:A375:zscore', 'BE4max:A375:zscore', 'BE39:OVCAR8:zscore',\n",
       "       'BE39:HAP1:zscore', 'BE39:HA1E:zscore', 'sgRNA_23bp'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_orig.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3872500d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 752/752 [00:09<00:00, 81.62it/s]\n"
     ]
    }
   ],
   "source": [
    "batch_size = 16\n",
    "features = []\n",
    "idx = -1\n",
    "for i in tqdm(range(0, len_data, batch_size)):\n",
    "    batch = data[i:i+batch_size]\n",
    "    seqs = [el[\"sequence\"] for el in batch]\n",
    "    inputs = tokenizer([f\"<|im_start|>{s}<|im_end|>\" for s in seqs], return_tensors=\"pt\", add_special_tokens=False, return_attention_mask=False)\n",
    "    inputs = {k: v.to(device) for k, v in inputs.items()}\n",
    "    with torch.inference_mode():\n",
    "        out = model(**inputs)\n",
    "\n",
    "    batch_features = out.hidden_states[:, idx]\n",
    "    batch_features = F.normalize(batch_features, p = 2, dim=1)\n",
    "    batch_features = batch_features.tolist()\n",
    "    features.extend(batch_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bbe4c475",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_ds = []\n",
    "for item, feature in zip(data, features):\n",
    "    new_item = {\n",
    "        **item,\n",
    "        **{f\"feature_{i}\": f for i, f in enumerate(feature)}\n",
    "    }\n",
    "    new_ds.append(new_item)\n",
    "\n",
    "final_ds = pd.DataFrame(new_ds)\n",
    "final_ds.to_csv(f\"/media/mascit/data/Projects/python/nucelotides_downstream/data.tmp/std_cls__ds20251123__exp20251118_0_lm_30m_idx{idx}_1.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da5c83e2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gradientlab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
